From 6f488b931f90132720f2dbf1476c15faa1e3e846 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Matthew=20=F0=9F=AA=B7?= <matthew@urgero.net>
Date: Sat, 14 Dec 2024 05:55:00 -0600
Subject: [PATCH 04/46] Winter Solstice 2024 Upgrade (Ruth epoch) (#5)

There are no changes that will activate with this upgrade.
---
 src/chainparams.cpp                           |  8 ++++
 src/consensus/activation.cpp                  |  9 ++++
 src/consensus/activation.h                    |  4 ++
 src/consensus/addresses.h                     |  4 ++
 src/consensus/params.h                        |  4 ++
 src/init.cpp                                  |  3 +-
 .../logos_feature_minerfund_activation.py     | 46 ++++++++++++++++++-
 7 files changed, 75 insertions(+), 3 deletions(-)

diff --git a/src/chainparams.cpp b/src/chainparams.cpp
index 60ceb6d5b..ea22c8cc7 100644
--- a/src/chainparams.cpp
+++ b/src/chainparams.cpp
@@ -117,6 +117,8 @@ public:
         consensus.judgesActivationTime = 1718916660;
         // 2024-12-21T09:20:00.000Z protocol upgrade
         consensus.ruthActivationTime = 1734772800;
+        // 2025-06-21T02:42:00.000Z protocol upgrade
+        consensus.firstSamuelActivationTime = 1750473720;
 
         /**
          * The message start string is designed to be unlikely to occur in
@@ -268,6 +270,10 @@ public:
         consensus.ruthActivationTime =
             mainnetConsensus.ruthActivationTime -
             testnetActivationOffset;
+        // 2025-06-21T02:42:00.000Z protocol upgrade
+        consensus.firstSamuelActivationTime =
+            mainnetConsensus.firstSamuelActivationTime -
+            testnetActivationOffset;
 
         // "ltdk" with MSB set
         diskMagic[0] = 0xec;
@@ -384,6 +390,8 @@ public:
         consensus.judgesActivationTime = mainnetConsensus.judgesActivationTime;
         // 2024-12-21T09:20:00.000Z protocol upgrade
         consensus.ruthActivationTime = mainnetConsensus.ruthActivationTime;
+        // 2025-06-21T02:42:00.000Z protocol upgrade
+        consensus.firstSamuelActivationTime = mainnetConsensus.firstSamuelActivationTime;
 
         // "lrdk" with MSB set
         diskMagic[0] = 0xec;
diff --git a/src/consensus/activation.cpp b/src/consensus/activation.cpp
index 4154e3729..c187a7608 100644
--- a/src/consensus/activation.cpp
+++ b/src/consensus/activation.cpp
@@ -72,3 +72,12 @@ bool IsJudgesEnabled(const Consensus::Params &params,
     return pindexPrev->GetMedianTimePast() >=
            gArgs.GetArg("-judgesactivationtime", params.judgesActivationTime);
 }
+
+bool IsRuthEnabled(const Consensus::Params &params,
+                          const CBlockIndex *pindexPrev) {
+    if (pindexPrev == nullptr) {
+        return false;
+    }
+    return pindexPrev->GetMedianTimePast() >=
+           gArgs.GetArg("-ruthactivationtime", params.ruthActivationTime);
+}
\ No newline at end of file
diff --git a/src/consensus/activation.h b/src/consensus/activation.h
index 2512ec813..68838ea64 100644
--- a/src/consensus/activation.h
+++ b/src/consensus/activation.h
@@ -39,4 +39,8 @@ bool IsJoshuaEnabled(const Consensus::Params &params,
 bool IsJudgesEnabled(const Consensus::Params &params,
                           const CBlockIndex *pindexPrev);
 
+/** Check if December 21st, 2024 protocol upgrade has activated. */
+bool IsRuthEnabled(const Consensus::Params &params,
+                          const CBlockIndex *pindexPrev);
+
 #endif // BITCOIN_CONSENSUS_ACTIVATION_H
diff --git a/src/consensus/addresses.h b/src/consensus/addresses.h
index 3de817010..cfa91d497 100644
--- a/src/consensus/addresses.h
+++ b/src/consensus/addresses.h
@@ -138,6 +138,10 @@ Consensus::CoinbaseAddresses AddressSets = {
             "lotus_16PSJJYp3YkVyctW8LRbp6UhbsrQMJghL3jjze3b7",
         },
     .judges =
+        {
+            "lotus_16PSJMaps9sQg7aBQgyY1RdHb2fZYdmWhQPbgus75"
+        },
+    .ruth =
         {
             "lotus_16PSJMaps9sQg7aBQgyY1RdHb2fZYdmWhQPbgus75"
         }
diff --git a/src/consensus/params.h b/src/consensus/params.h
index 42062d954..47fccd762 100644
--- a/src/consensus/params.h
+++ b/src/consensus/params.h
@@ -22,6 +22,7 @@ struct CoinbaseAddresses {
     std::vector<std::string> deuteronomy;
     std::vector<std::string> joshua;
     std::vector<std::string> judges;
+    std::vector<std::string> ruth;
 };
 
 /**
@@ -51,6 +52,9 @@ struct Params {
     /** Unit time used for MTP activation of 2024-12-21T09:20:00.000Z protocol
      * upgrade */
     int ruthActivationTime;
+    /** Unit time used for MTP activation of 2025-06-21T02:42:00.000Z protocol
+     * upgrade */
+    int firstSamuelActivationTime;
 
     /**
      * Don't warn about unknown BIP 9 activations below this height.
diff --git a/src/init.cpp b/src/init.cpp
index 1e048b2e5..082110d67 100644
--- a/src/init.cpp
+++ b/src/init.cpp
@@ -426,7 +426,8 @@ void SetupServerArgs(NodeContext &node) {
         // TODO remove after the December 2021 upgrade
         "-exodusactivationtime", "-leviticusactivationtime",
         "-numbersactivationtime", "-deuteronomyactivationtime",
-        "-joshuaactivationtime", "-judgesactivationtime"};
+        "-joshuaactivationtime", "-judgesactivationtime",
+        "-ruthactivationtime"};
 
     // Set all of the args and their help
     // When adding new options to the categories, please keep and ensure
diff --git a/test/functional/logos_feature_minerfund_activation.py b/test/functional/logos_feature_minerfund_activation.py
index a1bfe1a5a..967b58fd7 100755
--- a/test/functional/logos_feature_minerfund_activation.py
+++ b/test/functional/logos_feature_minerfund_activation.py
@@ -32,8 +32,9 @@ DEUTERONOMY_ACTIVATION_TIME = 2030000000
 JOSHUA_ACTIVATION_TIME = 2040000000
 JUDGES_ACTIVATION_TIME = 2050000000
 RUTH_ACTIVATION_TIME = 2060000000
+FIRST_SAMUEL_ACTIVATION_TIME = 2070000000
 
-REPLAYPROTECTION_ACTIVATION_TIME = RUTH_ACTIVATION_TIME
+REPLAYPROTECTION_ACTIVATION_TIME = FIRST_SAMUEL_ACTIVATION_TIME
 
 # see consensus/addresses.h, use getaddressinfo to get the scriptPubKey
 GENESIS_SCRIPTS = [
@@ -126,6 +127,9 @@ JOSHUA_SCRIPTS = [
 JUDGES_SCRIPTS = [
     "76a9149208ecc785c92968481a92aee7024b77e54d27dc88ac",
 ]
+RUTH_SCRIPTS = [
+    "76a9149208ecc785c92968481a92aee7024b77e54d27dc88ac",
+]
 
 class MinerFundActivationTest(BitcoinTestFramework):
     def set_test_params(self):
@@ -387,7 +391,45 @@ class MinerFundActivationTest(BitcoinTestFramework):
             block = make_block_cb_post_numbers(JUDGES_SCRIPTS)
             prepare_block(block)
             assert_equal(node.submitblock(ToHex(block)), None)
-
+        #
+        #   Begin Winter 2024 Upgrade Test
+        #
+        # Using the ruth addresses before upgrade fails
+        block = make_block_cb_post_numbers(RUTH_SCRIPTS)
+        prepare_block(block)
+        assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
+        node.setmocktime(JUDGES_ACTIVATION_TIME)
+        # Mine 11 blocks with RUTH_ACTIVATION in the middle
+        # That moves MTP exactly to RUTH_ACTIVATION
+        for i in range(-6, 6):
+            block = make_block_cb_post_numbers(JUDGES_SCRIPTS)
+            block.nTime = RUTH_ACTIVATION_TIME + i
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), None)
+        assert_equal(node.getblockchaininfo()['mediantime'],
+                     RUTH_ACTIVATION_TIME)
+        
+        # Now the using the genesis, exodus, leviticus, or numbers, and deuteronomy addresses fails
+        for block in [
+            make_block_with_cb_scripts(GENESIS_SCRIPTS),
+            make_block_with_cb_scripts(EXODUS_SCRIPTS),
+            make_block_with_cb_scripts(LEVITICUS_SCRIPTS),
+            make_block_cb_post_numbers(NUMBERS_SCRIPTS),
+            make_block_cb_post_numbers(DEUTERONOMY_SCRIPTS),
+            make_block_cb_post_numbers(JOSHUA_SCRIPTS),
+            make_block_cb_post_numbers(JUDGES_SCRIPTS),
+        ]:
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
+       
+        # Using the Ruth scripts now works
+        for i in range(0, 12):
+            block = make_block_cb_post_numbers(RUTH_SCRIPTS)
+            prepare_block(block)
+            assert_equal(node.submitblock(ToHex(block)), None)
+        #
+        #   End Winter 2024 Upgrade Test
+        #
         # Check replay protection is not enabled yet
         tx = CTransaction()
         tx.vin = [CTxIn(COutPoint(int(cointxid, 16), 1))]
-- 
2.39.5 (Apple Git-154)

